Forward and MFT track association to collisions

mcoquet642 opened the pull request:
@jgrosseo 
@sarahherrmann and I added a simple implementation of forward and mft track association to collisions, based on the code of `trackToCollisionAssociator.cxx`.

jgrosseo commented on Common/DataModel/CollisionAssociation.h:
```diff
@@ -33,6 +33,29 @@ DECLARE_SOA_TABLE(TrackAssoc, "AOD", "TRACKASSOC", //! Table for track-to-collis
+                  fwdtrack_association::CollisionId,
+                  fwdtrack_association::FwdTrackId);
+
+DECLARE_SOA_TABLE(FwdTrackCompColls, "AOD", "FWDTRACKCOLLS", //! Table with vectors of collision indices stored per fwdtrack
```
```suggestion
DECLARE_SOA_TABLE(FwdTrackCompColls, "AOD", "FWDTRACKCOMPCOLL", //! Table with vectors of collision indices stored per fwdtrack
```

jgrosseo commented on Common/DataModel/CollisionAssociation.h:
```diff
@@ -33,6 +33,29 @@ DECLARE_SOA_TABLE(TrackAssoc, "AOD", "TRACKASSOC", //! Table for track-to-collis
+                  fwdtrack_association::CollisionId,
+                  fwdtrack_association::MFTTrackId);
+
+DECLARE_SOA_TABLE(MFTTrackCompColls, "AOD", "MFTTRKCOMPCOLLS", //! Table with vectors of collision indices stored per mfttrack
```
```suggestion
DECLARE_SOA_TABLE(MFTTrackCompColls, "AOD", "MFTTRKCOMPCOLL", //! Table with vectors of collision indices stored per mfttrack
```

jgrosseo replied:
I guess you shortened it because it has a 14 (?) character limit. However, if first and third argument are not the same (except the s), then we can never write the table to a file which is probably a bad idea...

jgrosseo reviewed (commented):
Thanks for making this draft!

We have a lot of code duplication now (more or less the same code for MFT and forward, which itself is very similar to the one for central barrel tracks). We should find a way to template this, to avoid the duplication. Otherwise it will be very hard to maintain.

I also made 2 minor comments to make the table names canonical (actually there was a typo in the one for central barrel tracks, I just made a PR).

mcoquet642 commented:
Hi,
Thanks for the comments. Indeed we shortened the names because of the character limit. But I was not aware of the issue you pointed out. I tried to modify this in my last commit.

I also agree that it would be good to have a templated code for MFT and forward tracks. I will try to implement this and do another commit soon.

alibuild commented:
Error while checking build/O2Physics/o2 for 0670be28b70c4e2606a6e5c4d852ce2e3e2c2709 at 2023-05-02 19:40:
```
## sw/BUILD/O2Physics-latest/log
/sw/SOURCES/O2Physics/2525-slc7_x86-64/0/Common/TableProducer/fwdtrackToCollisionAssociator.cxx:32:12: error: 'FwdTrackCompColls' was not declared in this scope
/sw/SOURCES/O2Physics/2525-slc7_x86-64/0/Common/TableProducer/fwdtrackToCollisionAssociator.cxx:32:29: error: template argument 1 is invalid
/sw/SOURCES/O2Physics/2525-slc7_x86-64/0/Common/TableProducer/fwdtrackToCollisionAssociator.cxx:34:12: error: 'MFTTrackCompColls' was not declared in this scope
/sw/SOURCES/O2Physics/2525-slc7_x86-64/0/Common/TableProducer/fwdtrackToCollisionAssociator.cxx:34:29: error: template argument 1 is invalid
/sw/SOURCES/O2Physics/2525-slc7_x86-64/0/Common/TableProducer/fwdtrackToCollisionAssociator.cxx:118:28: error: expression cannot be used as a function
/sw/SOURCES/O2Physics/2525-slc7_x86-64/0/Common/TableProducer/fwdtrackToCollisionAssociator.cxx:120:28: error: expression cannot be used as a function
/sw/SOURCES/O2Physics/2525-slc7_x86-64/0/Common/TableProducer/fwdtrackToCollisionAssociator.cxx:146:28: error: expression cannot be used as a function
/sw/SOURCES/O2Physics/2525-slc7_x86-64/0/Common/TableProducer/fwdtrackToCollisionAssociator.cxx:148:28: error: expression cannot be used as a function
/sw/SOURCES/O2Physics/2525-slc7_x86-64/0/Common/TableProducer/fwdtrackToCollisionAssociator.cxx:227:28: error: expression cannot be used as a function
/sw/SOURCES/O2Physics/2525-slc7_x86-64/0/Common/TableProducer/fwdtrackToCollisionAssociator.cxx:229:28: error: expression cannot be used as a function
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/O2Physics/2525/0670be28b70c4e2606a6e5c4d852ce2e3e2c2709/build_O2Physics_o2/pretty.html).

mcoquet642 commented:
Hi @jgrosseo 
We fixed the previous error with the table names. We also implemented a version of the code using templated functions. Is this ok for you ?

jgrosseo commented:
@mcoquet642 Thanks for making a solution with templates, this is much better.

I don't want to delay the work further, so we can merge as it is. I was just wondering if the code would be much shorter if you wouldn't use the trick with the `constexpr BIT` but directly passed the arguments which are different. So for instance:

`runStandardAssoc<mftsPerCollisions, mftassociation, mftreverseIndices> (...)`

If I am not overlooking anything, then all `constexpr` would go away.

Similar for the association with time, where one would need to pass the `Id` column in addition.

Let me know if you would like to try this or not.

jgrosseo commented:
This is for instance used here: https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/Tasks/taskMcEfficiency.cxx#L599
(argument `pdgCodes`)

mcoquet642 commented:
Hi @jgrosseo , thanks for the suggestion. I tried to follow it, but I didn't manage to find the right way to pass the `Id` column as an argument of the time association, as you suggested. Would you have an idea how to do it ?

mcoquet642 commented on Common/TableProducer/fwdtrackToCollisionAssociator.cxx:
```diff
@@ -0,0 +1,201 @@
+      } else {
+        for (const auto& ambTrack : ambiguousTracks) {
+          if constexpr (static_cast<bool>(TTrackFillMap & BIT(0))) {
+            if (ambTrack.mfttrackId() == track.globalIndex()) {
```
```
 template <typename TTracks, typename TAmbiTracks, typename Assoc, typename RevIndices>
  void runAssocWithTime(Collisions const& collisions,
                        TTracks const& tracks,
                        TAmbiTracks const& ambiguousTracks,
                        BCs const& bcs,
                        Assoc association, RevIndices reverseIndices)
  {
    // cache globalBC
    std::vector<uint64_t> globalBC;
    for (const auto& track : tracks) {
      if (track.has_collision()) {
        globalBC.push_back(track.collision().bc().globalBC());
      } else {
        for (const auto& ambTrack : ambiguousTracks) {
            if (ambTrack.template getId<TTracks> == track.globalIndex()) {
              globalBC.push_back(ambTrack.bc().begin().globalBC());
              break;
            }
        }
      }
    }
[....]
```

jgrosseo replied:
Does it work if you do `ambTrack.template getId<TTracks>()` ?

mcoquet642 replied:
Yes indeed! Sorry about that ... I'll push the commit

jgrosseo commented:
Great. I approved.

@fgrosa You see that this is very similar to the central barrel one and we have a templated version now. Do you think it may be worth to unify this (in a further PR) with the central barrel one? E.g. by moving the template function to a header? I think in general two different executables makes sense.

fgrosa commented:
> Great. I approved.
> 
> @fgrosa You see that this is very similar to the central barrel one and we have a templated version now. Do you think it may be worth to unify this (in a further PR) with the central barrel one? E.g. by moving the template function to a header? I think in general two different executables makes sense.

Hi @jgrosseo, yes absolutely! Two executables make sense, but indeed it would be nice to move the common code in a header file. I will take care of it

jgrosseo commented:
@mcoquet642 

For the fun of it :)

![image](https://github.com/AliceO2Group/O2Physics/assets/25029837/93b98349-14c0-4fd7-a703-a456617f4a4c)

jgrosseo commented:
And that without having parsed O2 in chatGPT...

aalkin commented:
Because it can read errors better than us! If the left argument of the `==` is a function, while the right one is a number, it is indeed makes sense to call it with `()` 👍

mcoquet642 commented:
Incredible! I'll keep that in mind for next time :)
